src/DataFlowScene.cpp

```diff
@@ -0,0 +1,709 @@
+  bool moveNode(NodeIndex const& index, QPointF newLocation) override;
+
+  // convenience functions
+  QUuid addNode(QString const& typeID, QPointF const& location, QUuid const& uuid);
+  QUuid addNode(std::unique_ptr<NodeDataModel>&& model, QPointF const& location, QUuid const& uuid = QUuid::createUuid());
+  ConnectionID addConnection(Node* left, PortIndex leftIdx, Node* right, PortIndex rightIdx, TypeConverter converter);
+
+  using SharedConnection = std::shared_ptr<Connection>;
+  using UniqueNode       = std::unique_ptr<Node>;
+
+  std::unordered_map<ConnectionID, SharedConnection> _connections;
+  std::unordered_map<QUuid, UniqueNode>              _nodes;
+  std::shared_ptr<DataModelRegistry>                 _registry;
+};
+
+
+DataFlowScene::DataFlowScene(std::shared_ptr<DataModelRegistry> registry, QObject* parent) 
+  : FlowScene(new DataFlowModel(std::move(registry)), parent) {
+  _dataFlowModel = static_cast<DataFlowModel*>(model());
+}
+
+std::shared_ptr<Connection>
+DataFlowScene::
+createConnection(Node& nodeIn,
+                  PortIndex portIndexIn,
+                  Node& nodeOut,
+                  PortIndex portIndexOut,
+                  TypeConverter const& converter) {
+  
+  auto connid = _dataFlowModel->addConnection(&nodeOut, portIndexOut, &nodeIn, portIndexIn, converter);
+
+  return _dataFlowModel->_connections[connid];
+}
+
+std::shared_ptr<Connection>
+DataFlowScene::
+restoreConnection(QJsonObject const &connectionJson)
+{
+  QUuid nodeInId  = QUuid(connectionJson["in_id"].toString());
+  QUuid nodeOutId = QUuid(connectionJson["out_id"].toString());
+
+  PortIndex portIndexIn  = connectionJson["in_index"].toInt();
+  PortIndex portIndexOut = connectionJson["out_index"].toInt();
+
+
+  ConnectionID connId;
+  connId.lNodeID = nodeOutId;
+  connId.rNodeID = nodeInId;
+  
+  connId.lPortID = portIndexOut;
+  connId.rPortID = portIndexIn;
+  
+  if (!_dataFlowModel->addConnection(_dataFlowModel->nodeIndex(nodeOutId), connId.lPortID, _dataFlowModel->nodeIndex(nodeInId), connId.rPortID)) 
+    return nullptr;
+
+  return _dataFlowModel->_connections[connId];
+  
+}
+
+void
+DataFlowScene::
+deleteConnection(Connection& connection)
+{
+  auto deleted = _dataFlowModel->removeConnection(
+    _dataFlowModel->nodeIndex(connection.getNode(PortType::Out)->id()),
+    connection.getPortIndex(PortType::Out),
+    _dataFlowModel->nodeIndex(connection.getNode(PortType::In)->id()),
+    connection.getPortIndex(PortType::Out));
+  Q_ASSERT(deleted);
+}
+
+Node&
+DataFlowScene::
+createNode(std::unique_ptr<NodeDataModel> && dataModel)
+{
+  auto uid = _dataFlowModel->addNode(std::move(dataModel), {0.0, 0.0});
+
+  return *_dataFlowModel->_nodes[uid];
+}
+
+Node&
+DataFlowScene::
+restoreNode(QJsonObject const& nodeJson)
+{
+  QString modelName = nodeJson["model"].toObject()["name"].toString();
+
+  auto uid = QUuid(nodeJson["id"].toString());
+ 
+  _dataFlowModel->addNode(modelName, point, uid);
```

Comment:
What is `point` here? MinGW [errors out here](https://ci.appveyor.com/project/paceholder/nodeeditor/build/1.0.530/job/c26y36xc0n5kracv#L210); it can't tell what variable `point` is supposed to be